bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerMetadataSerDe.java

```diff
@@ -66,59 +76,70 @@
+        os.write(VERSION_KEY_BYTES);
+        os.write(String.valueOf(version).getBytes(UTF_8));
+        os.write(LINE_SPLITTER_BYTES);
     }
 
-    private byte[] serializeVersion1(LedgerMetadata metadata) {
-        StringBuilder s = new StringBuilder();
-        s.append(VERSION_KEY).append(FIELD_SPLITTER)
-            .append(metadata.getMetadataFormatVersion()).append(LINE_SPLITTER);
-        s.append(metadata.getWriteQuorumSize()).append(LINE_SPLITTER)
-            .append(metadata.getEnsembleSize()).append(LINE_SPLITTER).append(metadata.getLength());
+    private static int readHeader(InputStream is) throws IOException {
+        checkState(LINE_SPLITTER_BYTES.length == 1, "LINE_SPLITTER must be single byte");
 
-        for (Map.Entry<Long, ? extends List<BookieSocketAddress>> entry : metadata.getAllEnsembles().entrySet()) {
-            s.append(LINE_SPLITTER).append(entry.getKey());
-            for (BookieSocketAddress addr : entry.getValue()) {
-                s.append(FIELD_SPLITTER);
-                s.append(addr.toString());
+        for (int i = 0; i < VERSION_KEY_BYTES.length; i++) {
+            int b = is.read();
+            if (b < 0 || ((byte) b) != VERSION_KEY_BYTES[i]) {
+                throw new IOException("Ledger metadata header corrupt at index " + i);
             }
         }
-
-        if (metadata.getState() == State.IN_RECOVERY) {
-            s.append(LINE_SPLITTER).append(V1_IN_RECOVERY_ENTRY_ID)
-                .append(FIELD_SPLITTER).append(V1_CLOSED_TAG);
-        } else if (metadata.getState() == State.CLOSED) {
-            s.append(LINE_SPLITTER).append(metadata.getLastEntryId())
-                .append(FIELD_SPLITTER).append(V1_CLOSED_TAG);
-        } else {
-            checkArgument(metadata.getState() == State.OPEN,
-                          String.format("Unknown state %s for V1 serialization", metadata.getState()));
+        byte[] versionBuf = new byte[MAX_VERSION_DIGITS];
+        int i = 0;
+        while (i < MAX_VERSION_DIGITS) {
+            int b = is.read();
+            if (b == LINE_SPLITTER_BYTES[0]) {
+                String versionStr = new String(versionBuf, 0, i, UTF_8);
+                try {
+                    return Integer.parseInt(versionStr);
+                } catch (NumberFormatException nfe) {
+                    throw new IOException("Unable to parse version number from " + versionStr);
+                }
+            } else if (b < 0) {
+                break;
+            } else {
+                versionBuf[i++] = (byte) b;
+            }
         }
+        throw new IOException("Unable to find end of version number, metadata appears corrupt");
+    }
 
+    public byte[] serialize(LedgerMetadata metadata) throws IOException {
+        int formatVersion = Math.min(maxLedgerMetadataFormatVersion, metadata.getMetadataFormatVersion());
+        final byte[] serialized;
+        switch (formatVersion) {
+        case 3:
```

Comment:
can you define constants for these versions?

Reply:
done